### PR TITLE
feat: compare subscriptions for different dict formats

### LINKED ISSUE #137: Check if subscription already exists
I would like to avoid posting duplicate subscriptions. 
I've already written a working function for this:

```python
# Get existing subs
subscription_list = cb_client.get_subscription_list()
# Build a new one
subscription = Subscription(**my_data)
# Only post if it not exists already:
if not subscription_already_exists(subscription, existing_subscriptions=subscription_list):
    logger.info("Posting new subscription: %s", subscription.json(indent=2))
    cb_client.post_subscription(subscription=subscription)
```

The function `subscription_already_exists`:

```python
def subscription_already_exists(subscription, existing_subscriptions):
    sub = subscription.dict(exclude={"id"}).copy()
    for _existing_sub in existing_subscriptions:
        ex_sub = _existing_sub.dict(exclude={"id"}).copy()
        equals = sub_dict_is_equal(ex_sub, sub)
        if equals:
            return True
    return False  # If still here, it's not in the list


def sub_dict_is_equal(first, second):
    if first.keys() != second.keys():
        logger.warning(
            "Subscriptions contain a different set of fields. "
            "Only comparing to new fields of the new one."
        )
    for k, v in first.items():
        ex_value = second.get(k, None)
        if isinstance(v, dict):
            equal = sub_dict_is_equal(v, ex_value)
            if equal:
                continue
            else:
                return False
        if not v:
            ex_value = bool(ex_value)
            v = bool(v)
        if v != ex_value:
            return False
    return True

```

The nasty recursive comparison is due to filip and fiware interaction. Empty attributes in the notification dict are sometimes None and sometimes `[]`. Without this, the feature would be even shorter.

I would opt to move the function to `post_subcription` and add a kwarg `check_duplicate` or some better fitting name.

### PR DESCRIPTION
**Pull request by FWuellhorst**
Closes #137 .

@tstorek : I added the feature. I will write tests etc. once you approve the basic idea.

### CODE REVIEW

File: filip/clients/ngsi_v2/cb.py
Comment (djs0109) [inline filip/clients/ngsi_v2/cb.py:1764]: Should it be "if not _value_is_not_none(v) **or** not _value_is_not_none(ex_value)" here? Otherwise the none existing keys will not be ignored.

File: filip/clients/ngsi_v2/cb.py
Comment (djs0109) [inline filip/clients/ngsi_v2/cb.py:1750]: return bool(value)?

File: filip/clients/ngsi_v2/cb.py
Comment (FWuellhorst) [reply filip/clients/ngsi_v2/cb.py:1764]: Mhm, depends, or? If I have an old subscription with key `e = None`, I want the new key `e =  5` to be an update. Using `or` would lead to `None == 5 -> true` ?!

Comment (djs0109) [reply filip/clients/ngsi_v2/cb.py:1764]: That's true. A `None` normally indicates that the parameter is unset and the default value is used. So I think it would be ok if this function gives a warning or even returns these unequal keys as results? How to deal with these keys will then be the job of specific use case.

Comment (FWuellhorst) [reply filip/clients/ngsi_v2/cb.py:1764]: Ok! Do you want to add a warning yourself and make the changes? Or should I do that?

Comment (djs0109) [reply filip/clients/ngsi_v2/cb.py:1764]: I will do it ;)

**Comment by djs0109** [comment]
I think we still need a test for this new feature. @FWuellhorst Can you write a test for this new feature under `test_ngsi_v2_cb`? For example, the most basic test will be: try to post the same or slightly different `Subscription` twice.

**Comment by djs0109** [comment]
@FWuellhorst sry for the delay, can you still remember the reason to define another test subscription, i.e. `sub_with_nans`? This current cause some tiny issue

**Comment by FWuellhorst** [comment]
@djs0109 I think I had some subscription with nan values like `[]` which raised an issue. I can have a look at the logic and try to fix it, or do you want to?

**Comment by djs0109** [comment]
@FWuellhorst I fixed the empty expression issue. But there is still some tiny issue remaining caused by the for loop. It would be great if you can have a look :)

**Comment by tstorek** [comment]
@djs0109 @FWuellhorst There might be also tiny issue with the default values of the `Subscription` model. As @lidiawestphal notes in #181 in case that `attrs` is `none` or not included in the model and `explcitAttrs` is not present, Orion might assume an empty list for `attrs`. Hence, the server-subscription and the client subscription instantaneously are out of sync. That is also the reason why the `quick and dirty` variant with hasing the two json strings currently does not work. I am happy to help after my return. As registtrations need such a pattern as well :)

**Comment by djs0109** [comment]
The issue of `notification.attrs=None` is now implemented as a testcase, which can currently pass. So I assume that this issue is already solved. However, before this branch can be merged. I still want some extra tests to be implemented in `tests/clients/test_ngsi_v2_cb.py`

**Comment by djs0109** [comment]
Hi @RCX112, could you please complete this test here? I think after that we are ready to merge this PR
https://github.com/RWTH-EBC/FiLiP/blob/12789a0ede6dab99dfab2eadd9d91595048db9b3/tests/clients/test_ngsi_v2_cb.py#L637

**Comment by RCX112** [comment]
Hey @djs0109, I completed the test function in https://github.com/RWTH-EBC/FiLiP/blob/6155f430f5ad67a0382948dbb27bdc368c3937ec/tests/clients/test_ngsi_v2_cb.py#L534

where I added two extra subscriptions and also tested for duplicates. During the test I made some minor changes in the check function between two subscription dicts https://github.com/RWTH-EBC/FiLiP/blob/6155f430f5ad67a0382948dbb27bdc368c3937ec/filip/clients/ngsi_v2/cb.py#L1793

As an example if I have `condition1 = None` and `condition2 = {'attrs': [], 'expression': None}`, they should be both equal. I also ignored the attribute `timesSent`, because if the rest of the subscription dicts are equal I don't want to start counting from a second subscription, which is identical: https://github.com/RWTH-EBC/FiLiP/blob/6155f430f5ad67a0382948dbb27bdc368c3937ec/filip/clients/ngsi_v2/cb.py#L1841C1-L1845C29

**Comment by djs0109** [comment]
@RCX112 thank you, that looks good! I just slightly adjusted the code format. I will merge after the test run though

**Review by djs0109 — APPROVED** [review]

**Comment by FWuellhorst** [comment]
Thanks for finalizing this :)

**Pull request merged**
